src/permutations.jl

```diff
@@ -9,14 +9,50 @@ export
     permutations
 
 
-struct Permutations{T}
-    a::T
-    t::Int
+struct PermutationIterator{T}
+    data::T
+    length::Int
 end
 
-Base.eltype(::Type{Permutations{T}}) where {T} = Vector{eltype(T)}
+function has_repeats(state::Vector{Int})
+    for outer in eachindex(state)
+        for inner in (outer+1):lastindex(state)
+            if state[outer] == state[inner]
+                return true
+            end
+        end
+    end
+    return false
+end
+
+function increment!(state::Vector{Int}, min::Int, max::Int)
+    state[end] += 1
+    for i in reverse(eachindex(state))[firstindex(state):end-1]
+        if state[i] > max
+            state[i] = min
+            state[i-1] += 1
+        end
+    end
+end
+
+function next_permutation!(state::Vector{Int}, min::Int, max::Int)
+    while true
+        increment!(state, min, max)
+        has_repeats(state) || break
+    end
+end
+
+function Base.iterate(p::PermutationIterator, state::Vector{Int}=fill(firstindex(p.data), p.length))
+    next_permutation!(state, firstindex(p.data), lastindex(p.data))
+    if state[firstindex(state)] > lastindex(p.data)
+        return nothing
+    end
+    [p.data[i] for i in state], state
+end
```

Comment:
```suggestion
function Base.iterate(p::PermutationIterator, state::Vector{Int}=Int[])
    if isempty(state)
        # case for invalid t; iterator is empty
        p.length > length(p.data) && return nothing
        state = fill(firstindex(p.data), p.length)
    end
    # this is a special case if t=0 we return state [1] below which allows us to terminate early here
    length(state) == 1 && p.length == 0 && return nothing
    next_permutation!(state, firstindex(p.data), lastindex(p.data))
    if state[firstindex(state)] > lastindex(p.data)
        return nothing
    end
    # here you assume that p.data is not empty
    return [p.data[i] for i in state], p.length == 0 ? [1] : state
end
```

Reply:
Since `PermutationIterator` is not exported, is it ok if we do `@inferred <Vector{SomeType}> first(permutations(...))`

Or is there a clever way to use `PermutationIterator` in the tests without having to export it?

Reply:
Just create `PermutationIterator` manually in the tests and then call `iterate` on it.
The most important thing is to (apart from `@inferred`):
1. check if all works for empty collections and various `t`
2. check if all works for non-empty collections and various `t`
3. run performance tests of what I proposed against the earlier version (that did not use these workarounds - to make sure there is no regression; there might be some small regression, but not much)

Reply:
New benchmarks show some regression in timing, but not in allocations

Before PR
```julia
@benchmark foreach(identity, Combinatorics.permutations(1:2000, 2))
BenchmarkTools.Trial: 1 sample with 1 evaluation.
 Single result which took 10.268 s (30.58% GC) to evaluate,
 with a memory estimate of 60.47 GiB, over 11994001 allocations.
```

and after first attempt
```julia
@benchmark foreach(identity, Combinatorics.permutations(1:2000, 2))
BenchmarkTools.Trial: 39 samples with 1 evaluation.
 Range (min … max):  128.003 ms … 132.626 ms  ┊ GC (min … max): 8.80% … 9.58%
 Time  (median):     130.014 ms               ┊ GC (median):    9.60%
 Time  (mean ± σ):   130.239 ms ±   1.225 ms  ┊ GC (mean ± σ):  9.38% ± 0.41%

  █              █▃█   ▃  ▃  ▃              █   █                
  █▁▁▁▁▁▇▁▁▁▁▁▁▁▁███▁▇▁█▁▇█▇▇█▁▁▇▁▁▁▇▁▁▇▁▁▁▇█▇▁▁█▇▇▁▇▁▇▁▁▇▁▁▁▁▇ ▁
  128 ms           Histogram: frequency by time          133 ms <

 Memory estimate: 427.03 MiB, allocs estimate: 7996001.
```
With workarounds

```julia
@benchmark foreach(identity, Combinatorics.permutations(1:2000, 2))
BenchmarkTools.Trial: 31 samples with 1 evaluation.
 Range (min … max):  160.721 ms … 177.764 ms  ┊ GC (min … max): 14.08% … 16.64%
 Time  (median):     163.582 ms               ┊ GC (median):    15.21%
 Time  (mean ± σ):   164.714 ms ±   3.493 ms  ┊ GC (mean ± σ):  15.01% ±  0.74%

   ▁     ▁██▄  ▄          ▁                                      
  ▆█▁▆▁▆▁████▁▆█▁▆▁▁▆▆▁▁▁▁█▆▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  161 ms           Histogram: frequency by time          178 ms <

 Memory estimate: 427.03 MiB, allocs estimate: 7996002.
```
Looks like we are making the garbage collector do a little more work (9.38% vs. 15.01%)

Reply:
Yes - I expected regression due to additional checks that are required to preserve type stability.

@mschauer - what do you prefer:
1. type stable code that is a bit slower.
2. type unstable code that is faster?

(I would choose type unstable code, as it will be a small union - only two types, so compiler should be able to handle this efficiently)

@natemcintosh - I hope you are OK that we dig into details in this PR 😄.

Reply:
I would prefer the code that is simpler.

Reply:
About 1. I had to leave in a small part of your changes @bkamins because otherwise the test `@test collect(permutations(1:5, 1)) == [[x] for x in 1:5]` would fail. That left us with the following benchmark:

```julia
@benchmark foreach(identity, Combinatorics.permutations(1:2000, 2))
BenchmarkTools.Trial: 33 samples with 1 evaluation.
 Range (min … max):  151.879 ms … 166.773 ms  ┊ GC (min … max):  9.74% … 10.99%
 Time  (median):     154.526 ms               ┊ GC (median):    10.61%
 Time  (mean ± σ):   155.057 ms ±   2.564 ms  ┊ GC (mean ± σ):  10.46% ±  0.52%

  ▃    ▃█▃█▃▃   █▃▃ ▃  ▃                                         
  █▇▁▁▁██████▇▇▁███▁█▇▁█▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇ ▁
  152 ms           Histogram: frequency by time          167 ms <

 Memory estimate: 427.03 MiB, allocs estimate: 7996002.
```
Not quite back to initial time of ~130ms, but better than ~164ms.

Reply:
I think if you hit problems with `t=1` it would be better to have the fast code for `t>=2` and for `t=0` and `t=1` and `t<0` and `t>length(a)` have special cases. It would be both faster and the code would be simpler.

Reply:
To do this, my first thought would be to change the `permutations(a, t)` method to look like

```julia
function permutations(a, t::Integer)
    if (t <= 0) || (t > length(a))
        return Vector{eltype(a)}()
    elseif t == 1
        return ([ai] for ai in a)
    end
    return PermutationIterator(a, t)
end
```
PRO:
- The simplest code I could think of. But maybe there is simpler?
- The benchmark for `t=1` is pretty quick
```julia
@benchmark foreach(identity, permutations(1:2000, 1))
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  30.375 μs …  1.069 ms  ┊ GC (min … max): 0.00% … 96.95%
 Time  (median):     32.375 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   35.788 μs ± 50.855 μs  ┊ GC (mean ± σ):  7.38% ±  5.02%

       ▁█▅                                                     
  ▄▆▄▄▅███▇▅▄▃▄▄▄▅▄▃▄▄▃▄▅▄▄▄▄▅▅▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  30.4 μs         Histogram: frequency by time        41.8 μs <

 Memory estimate: 125.00 KiB, allocs estimate: 2000.
```

CON:
- Doing so would increase the size of the type union. E.g. for `a=1:5`:
```julia
Body::Union{Base.Generator{UnitRange{Int64}, Combinatorics.var"#70#71"}, Combinatorics.PermutationIterator{UnitRange{Int64}}, Vector{Int64}}
```
- The benchmark for `t=2` is basically the same:
```julia
@benchmark foreach(identity, permutations(1:2000, 2))
BenchmarkTools.Trial: 33 samples with 1 evaluation.
 Range (min … max):  144.337 ms … 160.524 ms  ┊ GC (min … max): 7.69% … 7.23%
 Time  (median):     156.691 ms               ┊ GC (median):    8.00%
 Time  (mean ± σ):   153.458 ms ±   5.712 ms  ┊ GC (mean ± σ):  8.19% ± 0.38%

                                                   █  ▅          
  █▁▁▁████▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▁▁▅█▅▅▅█▅▅█▁█▁▁▁▁▁▅ ▁
  144 ms           Histogram: frequency by time          161 ms <

 Memory estimate: 427.03 MiB, allocs estimate: 7996001.
```

Reply:
I put comments in the code
```
function permutations(a, t::Integer)
    if (t <= 0) || (t > length(a)) # case t=0 is different I think
        return Vector{eltype(a)}()
    elseif t == 1
        return ([ai] for ai in a) # you can produce a vector I think, or above produce a generator - so that the size of the union is 2
    end
    return PermutationIterator(a, t)
end
```

Reply:
Had a heavy homework workload last week. Just getting back to this now. 

So to summarize:
- for the case `t==0`, we want it to return a vector containing a single empty vector of the correct eltype
- for the cases `(t < 0) || (t > length(a))` and `t == 1`, have them both return the same type (either vector or generator)
    - I'll do some benchmarking and see which performs better

Reply:
> have them both return the same type
i.e. either the same type of vector as returned for `t==0` or generator of the same type as returned for other values of `t`

Reply:
Unfortunately it looks like the last push I made actually resulted in the slowest code yet, according to the small benchmark we've been using. 

```julia
julia> @benchmark foreach(identity, permutations(1:2000, 1))
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  33.708 μs …  1.723 ms  ┊ GC (min … max):  0.00% … 97.77%
 Time  (median):     36.791 μs              ┊ GC (median):     0.00%
 Time  (mean ± σ):   41.961 μs ± 91.404 μs  ┊ GC (mean ± σ):  12.14% ±  5.43%

                    ▆█▄                                        
  ▁▁▁▁▁▁▂▂▂▂▂▂▂▃▄▅▆▆████▅▃▃▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  33.7 μs         Histogram: frequency by time        43.1 μs <

 Memory estimate: 140.75 KiB, allocs estimate: 2001.

julia> @benchmark foreach(identity, permutations(1:2000, 2))
BenchmarkTools.Trial: 28 samples with 1 evaluation.
 Range (min … max):  179.495 ms … 187.204 ms  ┊ GC (min … max): 11.30% … 12.33%
 Time  (median):     184.790 ms               ┊ GC (median):    12.28%
 Time  (mean ± σ):   184.250 ms ±   1.475 ms  ┊ GC (mean ± σ):  12.04% ±  0.48%

                         ▁▁▁               ███▄                  
  ▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁███▁▁▁▁▆▁▁▁▁▁▆▁▁▁▆████▆▁▁▆▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  179 ms           Histogram: frequency by time          187 ms <

 Memory estimate: 427.03 MiB, allocs estimate: 7996001.
```
But the type union is at least smaller: for `permutations(1:3, 2)` `Union{Combinatorics.PermutationIterator{UnitRange{Int64}}, Vector{Vector{Int64}}}`

I guess this brings us back to the question: how much do we care about the size of the union as opposed to the speed of the code? Or as @mschauer said earlier, should we prioritize code that is as simple as possible?

Reply:
Given all these considerations - I would prioritize code simplicity. It is not likely that people will go back to this code often, so it is better to have it written in a way that is easy to understand.

Reply:
In that case, I think what we currently have is pretty simple. The new performance is still ~55.8x faster than the current state of `permutations`.

I have added just a few more comments to help clarity